modules/phase_field/src/userobjects/PolycrystalUserObjectBase.C

```diff
@@ -0,0 +1,284 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#include "PolycrystalUserObjectBase.h"
+#include "NonlinearSystemBase.h"
+#include "MooseMesh.h"
+#include "MooseVariable.h"
+
+#include "libmesh/dense_matrix.h"
+
+#include <vector>
+#include <map>
+#include <algorithm>
+
+template <>
+InputParameters
+validParams<PolycrystalUserObjectBase>()
+{
+  InputParameters params = validParams<FeatureFloodCount>();
+  params.addClassDescription("This object provides the base capability for creating proper reduced "
+                             "order parameter polycrystal initial conditions.");
+  params.addRequiredCoupledVarWithAutoBuild(
+      "variable", "var_name_base", "op_num", "Array of coupled variables");
+  params.addRequiredParam<unsigned int>(
+      "grain_num", "Number of grains being represented by the order parameters");
+  params.addParam<bool>("output_adjacency_matrix",
+                        false,
+                        "Output the Grain Adjacency Matrix used in the coloring algorithms. "
+                        "Additionally, the grain to OP assignments will be printed");
+  params.addParam<MooseEnum>("coloring_algorithm",
+                             PolycrystalUserObjectBase::coloringAlgorithms(),
+                             PolycrystalUserObjectBase::coloringAlgorithmDescriptions());
+
+  // Hide the output of the IC objects by default, it doesn't change over time
+  params.set<std::vector<OutputName>>("outputs") = {"none"};
+
+  // This object should only be executed _before_ the initial condition
+  MultiMooseEnum execute_options(SetupInterface::getExecuteOptions());
+  execute_options = "initial";
+  params.set<MultiMooseEnum>("execute_on") = execute_options;
+
+  return params;
+}
+
+PolycrystalUserObjectBase::PolycrystalUserObjectBase(const InputParameters & parameters)
+  : FeatureFloodCount(parameters),
+    _dim(_mesh.dimension()),
+    _op_num(_vars.size()),
+    _grain_num(getParam<unsigned int>("grain_num")),
+    _grain_to_op(_grain_num, PolycrystalUserObjectBase::INVALID_COLOR),
+    _coloring_algorithm(getParam<MooseEnum>("coloring_algorithm")),
+    _initialized(false),
+    _output_adjacency_matrix(getParam<bool>("output_adjacency_matrix"))
+{
+  if (_op_num > _grain_num)
+    mooseError("ERROR in PolycrystalVoronoi: Number of order parameters (op_num) can't be larger "
```

Comment:
This logic was copied over from the existing class. There's actually no reason to limit it other than it's dumb.